util/src/main/java/tc/oc/pgm/util/bukkit/Platform.java

```diff
@@ -2,62 +2,103 @@
 
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Supplier;
 import org.bukkit.Bukkit;
 import tc.oc.pgm.util.ClassLogger;
 import tc.oc.pgm.util.nms.NMSHacksNoOp;
 import tc.oc.pgm.util.nms.NMSHacksPlatform;
+import tc.oc.pgm.util.nms.v1_10_12.NMSHacks1_10_12;
+import tc.oc.pgm.util.nms.v1_8.NMSHacks1_8;
 import tc.oc.pgm.util.nms.v1_8.NMSHacksSportPaper;
 import tc.oc.pgm.util.nms.v1_9.NMSHacks1_9;
-import tc.oc.pgm.util.reflect.ReflectionUtils;
 
 public enum Platform {
   UNKNOWN("UNKNOWN", "UNKNOWN", NMSHacksNoOp.class, false),
   SPORTPAPER_1_8("SportPaper", "1.8", NMSHacksSportPaper.class, false),
-  SPIGOT_1_8(
-      "Spigot",
-      "1.8", // NMSHacks1_8 causes issues with other versions, get it dynamically
-      (Class<? extends NMSHacksPlatform>)
-          ReflectionUtils.getClassFromName("tc.oc.pgm.util.nms.v1_8.NMSHacks1_8"),
-      false),
-  PAPER_1_8(
-      "Paper",
-      "1.8", // NMSHacks1_8 causes issues with other versions, get it dynamically
-      (Class<? extends NMSHacksPlatform>)
-          ReflectionUtils.getClassFromName("tc.oc.pgm.util.nms.v1_8.NMSHacks1_8"),
-      false),
+  SPIGOT_1_8("Spigot", "1.8", () -> NMSHacks1_8.class, false),
+  PAPER_1_8("Paper", "1.8", () -> NMSHacks1_8.class, false),
   SPIGOT_1_9("Spigot", "1.9", NMSHacks1_9.class, true),
-  PAPER_1_9("Paper", "1.9", NMSHacks1_9.class, true);
+  PAPER_1_9("Paper", "1.9", NMSHacks1_9.class, true),
+  SPIGOT_1_10_12("Spigot", "1.10", "1.12", NMSHacks1_10_12.class, true),
+  PAPER_1_10_12("Paper", "1.10", "1.12", NMSHacks1_10_12.class, true);
 
   private static ClassLogger logger = ClassLogger.get(Platform.class);;
   public static Platform SERVER_PLATFORM = computeServerPlatform();
 
   private static Platform computeServerPlatform() {
     String versionString = Bukkit.getServer().getVersion();
     for (Platform platform : Platform.values()) {
-      if (versionString.contains(platform.variant)
-          && versionString.contains("MC: " + platform.majorVersion)) {
-        return platform;
+      for (String supportedMajorVersion : platform.getSupportedMajorVersions()) {
+        if (versionString.contains(platform.variant)
+            && versionString.contains("MC: " + supportedMajorVersion)) {
+          return platform;
+        }
       }
     }
     return UNKNOWN;
   }
 
   private final String variant;
-  private final String majorVersion;
-  private final Class<? extends NMSHacksPlatform> nmsHacksClass;
+  private final String majorVersionFirst;
+  private final String majorVersionLast;
```

Comment:
you can probably turn these into `Version fromVersion, toVersion`, and use PGM's utilities around versioning (constructor can still take in a string, and you can make it like 1.8.0 etc

Reply:
Presumably, I'd need to, for example with 1.9, have `"1.9.0", "1.9.4"`. I was hoping to continue ignoring the minor version/patch since the breaking changes so far all seem to be across major versions.